### ISSUE TITLE
Memory leak with 3.0.21 and recent 3.0.x

### ISSUE THREAD

**Issue by pokab** [FreeRADIUS/freeradius-server#3937]
I'm experiencing a really severe memleak. Tested 3.0.21 for a week, then compiled and tested the most recent version of 3.0.x branch.
OS: Ubuntu 20.04

Valgrind gives some useful output, run with very recent 3.0.x version:

```
==48759== 31,313,984 bytes in 128,336 blocks are possibly lost in loss record 907 of 909
==48759==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==48759==    by 0x4C34C86: _talloc_zero (in /usr/lib/x86_64-linux-gnu/libtalloc.so.2.3.0)
==48759==    by 0x4890153: dict_unknown_afrom_fields (in /usr/lib/freeradius/libfreeradius-radius.so)
==48759==    by 0x489CA94: data2vp (in /usr/lib/freeradius/libfreeradius-radius.so)
==48759==    by 0x489693B: fr_pair_make (in /usr/lib/freeradius/libfreeradius-radius.so)
==48759==    by 0x4896F6F: ??? (in /usr/lib/freeradius/libfreeradius-radius.so)
==48759==    by 0x14D9AB: ??? (in /usr/sbin/freeradius)
==48759==    by 0x14E257: ??? (in /usr/sbin/freeradius)
==48759==    by 0x4C51608: start_thread (pthread_create.c:477)
==48759==    by 0x4E77292: clone (clone.S:95)
==48759==
==48759== 56,893,968 bytes in 233,172 blocks are possibly lost in loss record 908 of 909
==48759==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==48759==    by 0x4C34C86: _talloc_zero (in /usr/lib/x86_64-linux-gnu/libtalloc.so.2.3.0)
==48759==    by 0x4890153: dict_unknown_afrom_fields (in /usr/lib/freeradius/libfreeradius-radius.so)
==48759==    by 0x489CA94: data2vp (in /usr/lib/freeradius/libfreeradius-radius.so)
==48759==    by 0x489693B: fr_pair_make (in /usr/lib/freeradius/libfreeradius-radius.so)
==48759==    by 0x4896F6F: ??? (in /usr/lib/freeradius/libfreeradius-radius.so)
==48759==    by 0x14D9AB: ??? (in /usr/sbin/freeradius)
==48759==    by 0x14E388: ??? (in /usr/sbin/freeradius)
==48759==    by 0x4C51608: start_thread (pthread_create.c:477)
==48759==    by 0x4E77292: clone (clone.S:95)
...
after restart:
==24412== 328,155,112 bytes in 1,344,898 blocks are still reachable in loss record 906 of 906
==24412==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==24412==    by 0x4C34C86: _talloc_zero (in /usr/lib/x86_64-linux-gnu/libtalloc.so.2.3.0)
==24412==    by 0x4890153: dict_unknown_afrom_fields (dict.c:2937)
==24412==    by 0x489CA94: data2vp (radius.c:3934)
==24412==    by 0x489693B: fr_pair_make_unknown (pair.c:1434)
==24412==    by 0x489693B: fr_pair_make (pair.c:1531)
==24412==    by 0x4896F6F: fr_pair_list_afrom_str.part.0 (pair.c:1959)
==24412==    by 0x14D9AB: detail_poll (detail.c:693)
==24412==    by 0x14E388: detail_handler_thread (detail.c:1115)
==24412==    by 0x4C51608: start_thread (pthread_create.c:477)
==24412==    by 0x4E77292: clone (clone.S:95)
```

Our configuration works as follows:
- it's an accounting-only server
- packets are dumped to disk using "detail" module for archival
- same packets are dumped to disk again, into a different directory for inserting into SQL later (let's call this a "work" directory)
- this "work" directory is read and processed using:
```
        listen {
                type = detail
                filename = "/mnt/<specific directory>/detail-*:*"
                load_factor = 80
        }
```
- we have 3 working sites with this exact setup

It seems that if I stop one specific ingesting instance (configured as descibed above), the memory leak is much slower or nonexistent. This instance is the only one which receives vendor-specific or other unknown attributes.
The attributes are, for example: `Attr-224`, `Attr-26.2011.168`, `Attr-26.29168.7`
Also, this instance sometimes stops processing the files from the "work" directory, and gets stuck. This only happened with the most recent build.

**Comment by alandekok** [comment]
I've pushed a fix which should help.  Please test and report if it works (or doesn't).

**Comment by pokab** [comment]
Will try, thanks.

**Comment by pokab** [comment]
Your newly added short circuit logic is not used very much, it seems:
```
   4122                 /*
   4123                  *      If it's already unknown, don't create a new
   4124                  *      unknown one.
   4125                  */
   4126                 if (da->flags.is_unknown) break;
```

It does not seem to help to avoid allocations:
```
==47770== 112,720,680 bytes in 461,970 blocks are still reachable in loss record 919 of 920
==47770==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==47770==    by 0x4C35C86: _talloc_zero (in /usr/lib/x86_64-linux-gnu/libtalloc.so.2.3.0)
==47770==    by 0x4890153: dict_unknown_afrom_fields (dict.c:2937)
==47770==    by 0x489CBBB: data2vp (radius.c:4133)
==47770==    by 0x489693B: fr_pair_make_unknown (pair.c:1434)
==47770==    by 0x489693B: fr_pair_make (pair.c:1531)
==47770==    by 0x4896F6F: fr_pair_list_afrom_str.part.0 (pair.c:1959)
==47770==    by 0x14D9AB: detail_poll (detail.c:693)
==47770==    by 0x14E257: detail_handler_thread (detail.c:1079)
==47770==    by 0x4C52608: start_thread (pthread_create.c:477)
==47770==    by 0x4E78292: clone (clone.S:95)
==47770==
==47770== 172,337,200 bytes in 706,300 blocks are still reachable in loss record 920 of 920
==47770==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==47770==    by 0x4C35C86: _talloc_zero (in /usr/lib/x86_64-linux-gnu/libtalloc.so.2.3.0)
==47770==    by 0x4890153: dict_unknown_afrom_fields (dict.c:2937)
==47770==    by 0x489CBBB: data2vp (radius.c:4133)
==47770==    by 0x489693B: fr_pair_make_unknown (pair.c:1434)
==47770==    by 0x489693B: fr_pair_make (pair.c:1531)
==47770==    by 0x4896F6F: fr_pair_list_afrom_str.part.0 (pair.c:1959)
==47770==    by 0x14D9AB: detail_poll (detail.c:693)
==47770==    by 0x14E388: detail_handler_thread (detail.c:1115)
==47770==    by 0x4C52608: start_thread (pthread_create.c:477)
==47770==    by 0x4E78292: clone (clone.S:95)
```

**Comment by alandekok** [comment]
Please attach a copy of a `detail` file entry which causes this issue.  Without that test case, this is a lot harder to fix.

**Comment by pokab** [comment]
I included two records in [detail.txt](https://github.com/FreeRADIUS/freeradius-server/files/6023160/detail.txt).
One is an Interim-Update, the other a Start record. They both have the possible culprits present.

Earlier I also made a memory dump using gdb and the /proc/.../smaps file of the process, as I continuously watched the virtual memory mappings used by the process. In those dumps, the most repeated readable ascii strings present were `Attr-26.2011.168` and `Attr-224`.

**Comment by alandekok** [comment]
I've tried that with the v3.0.x head, but I don't see any leaks.

**Comment by pokab** [comment]
Let me elaborate a bit.

According to the last valgrind output, "blocks are still reachable".
In the docs (https://www.valgrind.org/docs/manual/faq.html#faq.deflost) this is described as: "... means your program is probably ok -- it didn't free some memory it could have. This is quite common and often reasonable. Don't use --show-reachable=yes if you don't want to see these reports."

However, allocations of these reachable memory blocks seem to continue indefinitely (at least until 1 GB anonymous memory is exhausted). In contrast, version 2.x ran for 1 year at least, while consuming around 100 MB max.
This points to a bloated collection or cache, which still correctly accounts for each allocated block, but should not retain most of these elements, since they are irrelevant and are never used again. This in turn leads to (presumably) unbounded memory usage.

(This is a very common type of memory leak in garbage collected languages. Since the possibility of lost references is eliminated, this is what remains instead, which can be way more complicated to detect and fix.)

It there is any setting which limits the number of elements stored in any of these internal collections? Maybe the default is too high, or I set something too high without knowing how much memory usage that can cause.

**Comment by alandekok** [comment]
The code is written so that all data being allocated is put into various structures, which are then freed as appropriate.

So for every round of "read attributes from the packet and process them", the list of attributes _should_ be freed.

I've managed to verify this here, and push some fixes which will stop this from happening.  I'm also tracking down the underlying issue.

**Comment by pokab** [comment]
Please notify me when you are done with your changes, so that I can test them.

**Comment by alandekok** [comment]
I've verified that it's now fixed.

**Comment by pokab** [comment]
I have tried it, and immediately run into a possibly related problem.

The detail reader instance which was affected by the "unknown" fields effectively stops processing any records after a while. The "detail.work" file is open, the file pointer stops progressing and almost no CPU is used.
All other detail readers function normally. This makes me suspicious that this is also data-driven and related to the same root cause.

Also, it effectively makes me unable to run this version for a long enough time to verify any leaks, since leaks were caused by the affected detail reader progressing, which it does not do.

Is there a method to force the server to redo the work already partially done after restarting? It only continues processing the next detail file if I remove the file which is already there (detail.work), which in turn causes data loss. I would prefer that it starts processing the file from the beginning.

**Comment by alandekok** [comment]
You'll need to run the server in debug mode to see why it stops.  Odds are it's because the processing rules are not returning `ok`.  So the reader re-tries the same packet over and over.

The patches I pushed don't change anything about the processing.  They just stop the memory leak.

The default is for the server to process the detail.work file from the beginning.

Again, debug output would let you know exactly what's going on here.

I've verified here that the leak no longer occurs.
